Simpler definitions for standard gates

alexanderivrii opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

This PR slightly modernizes the definitions for standard gates:
* By using discrete gates instead of parametric gates whenever possible. For example, instances of ``U1(-pi/4)`` have been replaced by ``Tdg()``, instances of ``RZ(pi/2)`` have been replaced by ``S()`` with an update to the global phase, and so on.
*  The only definition that changed substantially is the one for ``ECRGate``, following [Jake's basis-constructor](https://github.com/Qiskit/qiskit-basis-constructor/blob/8c0bd17d2b92f2708f1329ba9d603fd41024315a/src/qiskit_basis_constructor/equivalence.py#L738-L743). In particular, ECR is now defined using ``1`` rather than ``4`` CX-gates.
* ``U1`` gates are replaced by ``Phase`` gates, ``U2`` and ``U3`` gates are replaced by ``U`` gates, ``CU1`` gates.

This is in the spirit of Jake's "complexity" classes for standard gates (Paulis < 1q Cliffords < 1q Cliffords + T/Tdg < parametric 1q gates). In particular, the definitions for Clifford standard gates are now in terms of simpler Clifford gates (e.g. for RC3X).

Update 1: Apparently we have CI tests that Rust-side definitions are essentially the same as Python-side definitions, so I have also changed the Python-side definitions.

Update 2: Apparently we have CI tests that Python-side definitions are included into the equivalence library, so some of the equivalences in the equivalence library are now updated (and in some cases a new equivalence is added).

### Details and comments

Question: can we eventually change the Python-side definitions to just return Rust-based definitions?

Question: are there potential problems with backward compatibility? Personally I think that we should be free to change definitions / synthesis methods and always aim to provide the best decompositions available, however what if there are users that expect that a definition for the ECRGate will contain a pair of RZX gates?

alexanderivrii committed:
ECR

alexanderivrii committed:
CRX

alexanderivrii committed:
CS

alexanderivrii committed:
CSdg

alexanderivrii committed:
CSX

alexanderivrii committed:
RYY

alexanderivrii committed:
XX-YY

alexanderivrii committed:
XX+YY

alexanderivrii committed:
RCCX

alexanderivrii committed:
RC3X

ShellyGarion commented:
Since U1Gate is going to be deprecated sometime, it's better to use PhaseGate instead of U1Gate and CPhaseGate instead of CU1Gate (these gates have exactly the same definition), see e.g.https://github.com/Qiskit/qiskit/issues/12977.

alexanderivrii committed:
Changing Python-side definitions as well.

We have explicit Rust tests that make sure that Python-side and Rust-side definitions are
identical.

alexanderivrii committed:
u1 -> phase, cu1 -> cphase, etc.

alexanderivrii committed:
lint

alexanderivrii committed:
fixing tests

alexanderivrii committed:
Bug fix

alexanderivrii committed:
undoing changes to a test

alexanderivrii commented on qiskit/circuit/_add_control.py:
```diff
@@ -170,9 +170,6 @@ def control(
 
             apply_basic_controlled_gate(controlled_circ, gate, q_control, target)
 
-            if gate.definition is not None and gate.definition.global_phase and gate.name != "rz":
-                global_phase += gate.definition.global_phase
```
This was an actual bug, which never got triggered because almost all of our equivalences do not update the global phase (and which got triggered after updating the definition for the ``ECRGate``).

alexanderivrii commented on crates/circuit/src/operations.rs:
```diff
@@ -1903,13 +1892,13 @@ impl Operation for StandardGate {
-                            (Self::RX, smallvec![Param::Float(-PI / 2.)], q0),
-                            (Self::RX, smallvec![Param::Float(-PI / 2.)], q1),
+                            (Self::SX, smallvec![], q0),
+                            (Self::SX, smallvec![], q1),
```
For a very attentive reviewer: RX and SXdg gates differ by a global phase, however all the 4 global phase contributions cancel each other out.

alexanderivrii commented on test/python/circuit/test_rust_equivalence.py:
```diff
@@ -74,63 +74,12 @@ def test_definitions(self):
-                                [py_def.find_bit(x).index for x in py_inst.qubits],
-                                [rs_def.find_bit(x).index for x in rs_inst.qubits],
-                            )
+                        self.assertEqual(py_inst.operation.name, rs_inst.operation.name)
+                        self.assertEqual(rs_inst.operation.params, py_inst.operation.params)
+                        self.assertEqual(
+                            [py_def.find_bit(x).index for x in py_inst.qubits],
+                            [rs_def.find_bit(x).index for x in rs_inst.qubits],
+                        )
```
Now that the Python-side definitions and Rust-side definitions are exactly identical, we do not need the special cases above.

alexanderivrii committed:
remove stray prints and blank lines

alexanderivrii commented on test/python/qasm3/test_export.py:
```diff
@@ -430,14 +430,19 @@ def test_multiple_pauli_evolution_gates(self):
+  h _gate_q_0;
+  s _gate_q_0;
+}
 gate ryy(p0) _gate_q_0, _gate_q_1 {
-  rx(pi/2) _gate_q_0;
-  rx(pi/2) _gate_q_1;
+  sxdg _gate_q_0;
+  sxdg _gate_q_1;
   cx _gate_q_0, _gate_q_1;
   rz(p0) _gate_q_1;
   cx _gate_q_0, _gate_q_1;
-  rx(-pi/2) _gate_q_0;
-  rx(-pi/2) _gate_q_1;
+  sx _gate_q_0;
+  sx _gate_q_1;
```
@Cryoris, I am wondering if the change for the standard definition of the RYYGate may have undesired/unexpected effects for pauli evolution synthesis, what do you think?

Cryoris replied:
Maybe some of the tests need adjustments, but the Pauli evolution code also uses SX/SXdg as Clifford basis changes for multi-qubit rotations. So if anything this makes is more consistent.

alexanderivrii committed:
reno

alexanderivrii committed:
docs

alexanderivrii committed:
Replacing Python definition by Rust definition; fixing problematic definitions for SX and SXdg in Rust

alexanderivrii committed:
updating python definitions

We can have a subclass of say XGate that sets _standard_gate = None, yet we
expect the definition of XGate

alexanderivrii committed:
Adding visualization for definition circuits

alexanderivrii commented on crates/circuit/src/operations.rs:
```diff
@@ -1338,7 +1338,7 @@ impl Operation for StandardGate {
                             (Self::Sdg, smallvec![], smallvec![Qubit(0)]),
                         ],
-                        FLOAT_ZERO,
+                        Param::Float(PI / 4.),
```
This global phase (and the global phase for SXdg decomposition) were incorrect in the Rust code here, but correct both in the Python code and in the equivalence library. Does it make sense to submit just these fixes in a separate PR so that we could backport it? I guess the only way the incorrect behavior could come up in practice is when calling HighLevelSynthesis (since it now uses Rust-based gate definitions) but without the standard equivalence library (since this beats definitions priority-wise).

ShellyGarion replied:
it seems as a good idea to me

ShellyGarion commented on qiskit/circuit/library/standard_gates/r.py:
```diff
@@ -62,22 +61,17 @@ def __init__(
+
+        #    ┌──────────────────────────┐
+        # q: ┤ U(θ,φ - π/2,π/2 - φ) ├
+        #    └──────────────────────────┘
```
the picture does not look well

alexanderivrii replied:
that was fixed

ShellyGarion commented on qiskit/circuit/library/standard_gates/ecr.py:
```diff
@@ -39,11 +37,13 @@ class ECRGate(SingletonGate):
-        q_0: ┤0        ├       q_0: ┤0           ├┤ RX(pi) ├┤0            ├
-             │   ECR   │   =        │  RZX(pi/4) │└────────┘│  RZX(-pi/4) │
-        q_1: ┤1        ├       q_1: ┤1           ├──────────┤1            ├
-             └─────────┘            └────────────┘          └─────────────┘
```
I see that this equivalence is also in the equivalence library. what happens if one decompose an ECR gate into a basis set that contains RZX or RZZ gates?

alexanderivrii replied:
We should be completely fine making changes if the equivalence library contains the old and the new definitions for all of the gates. We should also make sure that we do not remove definitions (and just to be sure I have just now restored one definition that I have removed).

ShellyGarion commented on qiskit/circuit/library/standard_gates/rx.py:
```diff
@@ -224,38 +220,18 @@ def __init__(
+        # q_0: ───────■───────────────────■───────────────────
+        #      ┌───┐┌─┴─┐┌─────────────┐┌─┴─┐┌───────────────┐
+        # q_1: ┤ S ├┤ X ├┤ U(-θ/2,0,0) ├┤ X ├┤ U(θ/2,-π/2,0) ├
+        #      └───┘└───┘└─────────────┘└───┘└───────────────┘
```
perhaps this definition could be simplified? I will look into it

ShellyGarion replied:
the equivalence library also has this definition:

```
# CRXGate
#
# q_0: ────■────     q_0: ───────■────────────────■────────────────────
#      ┌───┴───┐  ≡       ┌───┐┌─┴─┐┌──────────┐┌─┴─┐┌─────────┐┌─────┐
# q_1: ┤ Rx(ϴ) ├     q_1: ┤ S ├┤ X ├┤ Ry(-ϴ/2) ├┤ X ├┤ Ry(ϴ/2) ├┤ Sdg ├
#      └───────┘          └───┘└───┘└──────────┘└───┘└─────────┘└─────┘
```

alexanderivrii replied:
Good point, this definition is a bit simpler and a bit more powerful than the other one (in the sense that is uses single-parameter RY-gates instead of the more general U-gates). I will update the (default) definition.

ShellyGarion replied:
In the above picture it still uses U and not Ry (but the definition should be correct now?)

ShellyGarion commented on qiskit/circuit/library/standard_gates/x.py:
```diff
@@ -606,65 +554,13 @@ def __init__(
-        self.definition = qc
+        self.definition = QuantumCircuit._from_circuit_data(
+            StandardGate.C3SX._get_definition(self.params), add_regs=True, name=self.name
+        )
```
maybe add an image of the circuit?

alexanderivrii replied:
The definitions for the 3-controlled circuits are quite large and don't git on a single page. This is why I have provided the definitions for almost all circuits but not the very large ones.

ShellyGarion commented on qiskit/circuit/library/standard_gates/x.py:
```diff
@@ -701,76 +597,13 @@ def __init__(
-        self.definition = qc
+        self.definition = QuantumCircuit._from_circuit_data(
+            StandardGate.C3X._get_definition(self.params), add_regs=True, name=self.name
+        )
```
maybe add an image of the circuit?

alexanderivrii replied:
Please see the answer above.

ShellyGarion commented on qiskit/circuit/library/standard_gates/x.py:
```diff
@@ -871,59 +704,13 @@ def __init__(self, label: Optional[str] = None):
-        self.definition = qc
+        self.definition = QuantumCircuit._from_circuit_data(
+            StandardGate.RC3X._get_definition(self.params), add_regs=True, name=self.name
+        )
```
maybe add an image of the circuit?

alexanderivrii replied:
Please see the answer above.

ShellyGarion commented on qiskit/circuit/library/standard_gates/x.py:
```diff
@@ -963,46 +750,22 @@ def __init__(
         from .h import HGate
+        from .p import CPhaseGate
 
+        # This is not a standard gate in Rust
```
maybe add an image of the circuit?

alexanderivrii replied:
Please see the answer above.

ShellyGarion reviewed (commented):
Looks great! thanks for doing all this effort!
I have a few minor comments.

alexanderivrii committed:
Merge branch 'main' into simpler-defs

alexanderivrii committed:
addressing review comments

alexanderivrii committed:
improving release notes

alexanderivrii commented:
Thanks @ShellyGarion, I have addressed your comments modulo adding images for circuits with many gates: such circuits do not nicely fit on one screen and (imho) are not very illuminating. Can you please take another look?

ShellyGarion commented on releasenotes/notes/improve-standard-gate-definitions-1be3d27996cf15b6.yaml:
```diff
@@ -0,0 +1,14 @@
+    * When available, a definition using Clifford gates is preferred over
+      one that includes non-Clifford gates.
+    * The use of ``PhaseGate`` is preferred over ``U1Gate``.
+    * The use of ``UGate`` is preferred over ``U3Gate``.
```
also over U2Gate, right?

ShellyGarion replied:
Also using T/Tdg is preferable over PhaseGate

ShellyGarion reviewed (commented):
Amazing work! A few last comments and I think it's ready to go.

alexanderivrii committed:
Merge branch 'main' into simpler-defs

alexanderivrii committed:
Changing the picture for RCX gate to be consistent with the new definition

alexanderivrii committed:
improving release notes

ShellyGarion reviewed (approved):
LGTM. tnx!
